time-series-analyzer-template/src/agent.py

```diff
@@ -0,0 +1,208 @@
+        return FindingSeverity.Info
+    if finding_severity == "CRITICAL":
+        return FindingSeverity.Critical
+    if finding_severity == "HIGH":
+        return FindingSeverity.High
+    if finding_severity == "MEDIUM":
+        return FindingSeverity.Medium
+    if finding_severity == "LOW":
+        return FindingSeverity.Low
+
+    return FindingSeverity.Unknown
+
+
+def update_alerted_timestamp(timestamp: datetime):
+    """
+    this function maintains a time stamps; holds up to TIMESTAMP_QUEUE_SIZE in memory
+    :return: None
+    """
+    global ALERTED_TIMESTAMP
+
+    ALERTED_TIMESTAMP.append(timestamp)
+    if len(ALERTED_TIMESTAMP) > TIMESTAMP_QUEUE_SIZE:
+        ALERTED_TIMESTAMP.pop(0)
+
+
+def detect_attack(w3, forta_explorer, block_event: forta_agent.block_event.BlockEvent):
+    """
+    this function returns finding for any alert frequency for the most recent BUCKET that breaks out the predicted range by the Prophet time series model.
+    :return: findings: list
+    """
+    global ALERTED_TIMESTAMP
+    global FINDINGS_CACHE
+    global MUTEX
+
+    if not MUTEX:
+        MUTEX = True
+
+        # get time for block to derive date range for query
+        end_date = datetime.utcfromtimestamp(block_event.block.timestamp)
+        start_date = end_date - timedelta(minutes=BUCKET_WINDOW_IN_MINUTES * TRAINING_WINDOW_IN_BUCKET_SIZE)
+        logging.info(f"Analyzing alerts from {start_date} to {end_date}")
+
+        # get all alerts for date range
+        df_bot_alerts = forta_explorer.alerts_by_bot(BOT_ID, ALERT_NAME, CONTRACT_ADDRESS, start_date, end_date)
+        logging.info(f"Fetched {len(df_bot_alerts)} for bot_id {BOT_ID}, alert_id {ALERT_NAME}, contract_address {CONTRACT_ADDRESS}")
+
+        if len(df_bot_alerts) == 0:
+            logging.info("No alerts found for bot_id {BOT_ID}, alert_id {ALERT_NAME}, contract_address {CONTRACT_ADDRESS}")
+            MUTEX = False
+            return
+
+        # build time series model without last bucket
+        df_timeseries = df_bot_alerts.resample(str(BUCKET_WINDOW_IN_MINUTES) + 'min', on='createdAt').count()["hash"].reset_index()
+        df_timeseries['createdAt'] = df_timeseries['createdAt'].dt.tz_localize(None)
+
+        if len(df_timeseries) < 3:
+            logging.info("Not enough data to train model")
+            MUTEX = False
+            return
+
+        df_timeseries = df_timeseries[df_timeseries["createdAt"] < df_timeseries["createdAt"].max()]  # this row could be incomplete, so we discard
+        df_current_value = df_timeseries[df_timeseries["createdAt"] == df_timeseries["createdAt"].max()]
+        df_timeseries = df_timeseries[df_timeseries["createdAt"] < df_timeseries["createdAt"].max()]  # this row is what we want to assess against the model, so we discard
+
+        df_timeseries.rename(columns={'createdAt': 'ds', 'hash': 'y'}, inplace=True)
+        df_timeseries['ds'] = df_timeseries['ds'].dt.tz_localize(None)
```

Comment:
this can be removed because it's been localized on line 119. What does it do of `None` is passed in? Does it update it to UTC?

Reply:
it makes the time zone consistent, so the data can be joined; otherwise the concat of the df fails.